File: tools/SparseSolver/SparseSolver.cpp
/* This file is part of PyMesh. Copyright (c) 2018 by Qingnan Zhou */
#include "SparseSolver.h"
#include <sstream>

#include <Eigen/IterativeLinearSolvers>
#include <Eigen/SparseCholesky>
#include <Eigen/SparseLU>
#include <Eigen/SparseQR>
#ifdef WITH_UMFPACK
#include <Eigen/UmfPackSupport>
#endif
#ifdef WITH_CHOLMOD
#include <Eigen/CholmodSupport>
#endif
#ifdef WITH_MKL
#include <Eigen/PardisoSupport>
#endif

#include <Core/Exception.h>

#include "SparseSolverImplementation.h"

using namespace PyMesh;

SparseSolver::Ptr SparseSolver::create(const std::string& solver_type) {
    if (solver_type == "LDLT") {
        using LDLT = Eigen::SimplicialLDLT<ZSparseMatrix::ParentType>;
        return SparseSolver::Ptr(new SparseSolverImplementation<LDLT>);
    } else if (solver_type == "LLT") {
        using LLT = Eigen::SimplicialLLT<ZSparseMatrix::ParentType>;
        return SparseSolver::Ptr(new SparseSolverImplementation<LLT>);
    } else if (solver_type == "CG") {
        using CG = Eigen::ConjugateGradient<ZSparseMatrix::ParentType,
              Eigen::Lower|Eigen::Upper>;
        return SparseSolver::Ptr(
                new IterativeSparseSolverImplementation<CG>);
    } else if (solver_type == "LSCG") {
        using LSCG = Eigen::LeastSquaresConjugateGradient<ZSparseMatrix::ParentType>;
        return SparseSolver::Ptr(
                new IterativeSparseSolverImplementation<LSCG>);
    } else if (solver_type == "BiCG") {
        using BiCGSTAB = Eigen::BiCGSTAB<ZSparseMatrix::ParentType>;
        return SparseSolver::Ptr(
                new IterativeSparseSolverImplementation<BiCGSTAB>);
    } else if (solver_type == "SparseLU") {
        using SparseLU = Eigen::SparseLU<ZSparseMatrix::ParentType,
              Eigen::COLAMDOrdering<ZSparseMatrix::ParentType::StorageIndex> >;
        return SparseSolver::Ptr(new SparseSolverImplementation<SparseLU>);
    } else if (solver_type == "SparseQR") {
        using SparseQR = Eigen::SparseQR<ZSparseMatrix::ParentType,
              Eigen::COLAMDOrdering<ZSparseMatrix::ParentType::StorageIndex> >;
        return SparseSolver::Ptr(new SparseSolverImplementation<SparseQR>);
    }
#ifdef WITH_UMFPACK
    if (solver_type == "UmfPackLU") {
        using UmfPackLU = Eigen::UmfPackLU<ZSparseMatrix::ParentType>;
        return SparseSolver::Ptr(
                new SparseSolverImplementationWithLocalCopy<UmfPackLU>);
    }
#endif
#ifdef WITH_CHOLMOD
    if (solver_type == "Cholmod") {
        using CholmodSupernodalLLT =
            Eigen::CholmodSupernodalLLT<ZSparseMatrix::ParentType>;
        return SparseSolver::Ptr(
                new SparseSolverImplementation<CholmodSupernodalLLT>);
    }
#endif
#ifdef WITH_MKL
    if (solver_type == "PardisoLLT") {
        using PardisoLLT = Eigen::PardisoLLT<ZSparseMatrix::ParentType>;
        return SparseSolver::Ptr(
                new SparseSolverImplementation<PardisoLLT>);
    } else if (solver_type == "PardisoLDLT") {
        using PardisoLDLT = Eigen::PardisoLDLT<ZSparseMatrix::ParentType>;
        return SparseSolver::Ptr(
                new SparseSolverImplementation<PardisoLDLT>);
    } else if (solver_type == "PardisoLU") {
        using PardisoLU = Eigen::PardisoLU<ZSparseMatrix::ParentType>;
        return SparseSolver::Ptr(
                new SparseSolverImplementation<PardisoLU>);
    }
#endif

    std::stringstream err_msg;
    err_msg << "Unsupported solver type " << solver_type;
    throw NotImplementedError(err_msg.str());
}

std::vector<std::string> SparseSolver::get_supported_solvers() {
    std::vector<std::string> solver_names{
        "LDLT",
        "LLT",
        "CG",
        "LSCG",
        "BiCG",
        "SparseLU",
        "SparseQR"
    };
#ifdef WITH_UMFPACK
    solver_names.push_back("UmfPackLU");
#endif
#ifdef WITH_CHOLMOD
    solver_names.push_back("Cholmod");
#endif
#ifdef WITH_MKL
    solver_names.push_back("PardisoLLT");
    solver_names.push_back("PardisoLDLT");
    solver_names.push_back("PardisoLU");
#endif
    return solver_names;
}

File: python/pymesh/SparseSolver.py
from PyMesh import SparseSolver
import PyMesh

class SparseSolver(PyMesh.SparseSolver):
    """ Linear solver for solving sparse linear systems.

    This class is a thin wrapper around `sparse solvers supported by Eigen
    <https://eigen.tuxfamily.org/dox/group__TopicSparseSystems.html>`_.

    The following direct solvers are supported:

    * ``LLT``: Wrapper of `Eigen::SimplicialLLT`_, SPD only.
    * ``LDLT``: Wrapper of `Eigen::SimplicialLDLT`_, SPD only.
    * ``SparseLU``: Wrapper of `Eigen::SparseLU`_.
    * ``SparseQR``: Wrapper of `Eigen::SparseQR`_.
    * ``UmfPackLU``: Wrapper of `Eigen::UmfPackLU`_.  (Require `SuiteSparse`_)
    * ``Cholmod``: Wrapper of `Eigen::CholmodSupernodalLLT`_. (Require `SuiteSparse`_)
    * ``PardisoLLT``: Wrapper of `Eigen::PardisoLLT`_. SPD only. (Require `Intel MKL`_)
    * ``PardisoLDLT``: Wrapper of `Eigen::PardisoLDLT`_. SPD only. (Require `Intel MKL`_)
    * ``PardisoLU``: Wrapper of `Eigen::PardisoLU`_. (Require `Intel MKL`_)

    The following iterative solvers are supported:

    * ``CG``: Wrapper of `Eigen::ConjugateGradient`_. SPD only.
    * ``LSCG``: Wrapper of `Eigen::LeastSquaresConjugateGradient`_.
    * ``BiCG``: Wrapper of `Eigen::BiCGSTAB`_.

    Attributes:

        supported_solvers (``list`` of ``str``): The list of supported solvers.

        tolerance (``float``): The residual error threshold for stopping
            iterative solvers.  Default is ``Eigen::NumTraits<Float>::epsilon()``.

        max_iterations (``int``):  The max iterations allowed for iterative
            solvers.  Default is twice the number of columns of the matrix.

    Example:

        For direct solvers:

        >>> M = scipy.sparse.eye(100) # Build matrix.
        >>> rhs = numpy.ones(100) # build right hand side.
        >>> solver = pymesh.SparseSolver.create("LDLT")
        >>> solver.compute(M)
        >>> x = solver.solve(rhs)

        For iterative solvers:

        >>> M = scipy.sparse.eye(100) # Build matrix.
        >>> rhs = numpy.ones(100) # build right hand side.
        >>> solver = pymesh.SparseSolver.create("CG")
        >>> solver.tolerance = 1e-10
        >>> solver.max_iterations = 50
        >>> solver.compute(M)
        >>> x = solver.solve(rhs)

    .. _`Eigen::SimplicialLLT`: https://eigen.tuxfamily.org/dox/classEigen_1_1SimplicialLLT.html
    .. _`Eigen::SimplicialLDLT`: https://eigen.tuxfamily.org/dox/classEigen_1_1SimplicialLDLT.html
    .. _`Eigen::SparseLU`: https://eigen.tuxfamily.org/dox/classEigen_1_1SparseLU.html
    .. _`Eigen::SparseQR`: https://eigen.tuxfamily.org/dox/classEigen_1_1SparseQR.html
    .. _`Eigen::UmfPackLU`: https://eigen.tuxfamily.org/dox/classEigen_1_1UmfPackLU.html
    .. _`Eigen::CholmodSupernodalLLT`: https://eigen.tuxfamily.org/dox/classEigen_1_1CholmodSupernodalLLT.html
    .. _`Eigen::PardisoLLT`: https://eigen.tuxfamily.org/dox/classEigen_1_1PardisoLLT.html
    .. _`Eigen::PardisoLDLT`: https://eigen.tuxfamily.org/dox/classEigen_1_1PardisoLDLT.html
    .. _`Eigen::PardisoLU`: https://eigen.tuxfamily.org/dox/classEigen_1_1PardisoLU.html
    .. _`Eigen::ConjugateGradient`: https://eigen.tuxfamily.org/dox/classEigen_1_1ConjugateGradient.html
    .. _`Eigen::LeastSquaresConjugateGradient`: https://eigen.tuxfamily.org/dox/classEigen_1_1LeastSquaresConjugateGradient.html
    .. _`Eigen::BiCGSTAB`: https://eigen.tuxfamily.org/dox/classEigen_1_1BiCGSTAB.html
    .. _`SuiteSparse`: http://faculty.cse.tamu.edu/davis/suitesparse.html
    .. _`Intel MKL`: https://software.intel.com/en-us/mkl
    """
    @property
    def supported_solvers(self):
        return self.get_supported_solvers()

    @classmethod
    def get_supported_solvers(cls):
        return PyMesh.SparseSolver.get_supported_solvers()
